Kbuild: adding C++ support?

I'm testing the new Kbuild-based build system and trying to convert an app of our own to the new build system. One issue that I can't seem to work around is that some of the files in the app are written in C++ because of historical reasons. With the previous build system, I was able to add support for C++ files to the makefile simply by adding extra rules to the app-level Makefile like this:

```
CXX = $(CROSS_COMPILE)g++
LD = $(CROSS_COMPILE)g++

CXX_COMMAND=$(CXX) $(CFLAGS) -c $< -o $(BIN)/$@
CXX_COMMAND_SILENT="  CXX   $@"
.cpp.o:
     @$(if $(QUIET), ,echo $(CXX_COMMAND$(VERBOSE)) )
     @$(CXX_COMMAND)
```

It seems like this is not possible with the new Kbuild system because the app-level makefile eventually ends up invoking another `make` process with the top-level makefile, which does not have support for C++ files. In the previous build system, the top-level makefile was included instead of forking off to another `make` process so it was easy to add the modifications.

Would it be possible to add support for C++ source files directly in `tools/make/targets.mk` even if the current Crazyflie firmware does not use any C++ sources?

`-lstdc++` gets rid of a few of the undefined references but not all of them. I could not figure out how to add `-lstdc++` in an app-level `Kbuild` file (`libs-y += -lstdc++` seemed to have no effect) so I just added it to `../tools/kbuild/Makefile.kbuild` straight before `-lm` and then it was picked up properly.

The next problem was a missing reference to `abort()`. `abort()` is called implicitly by the `new` operator in C++ in case the allocation fails. I solved it by adding `obj-y += abort.o` to `src/utils/src/Kbuild`.

Next up: undefined references to `_write`, `_close`, `_fstat` and many others, none of which were called directly by our C++ code. It took me a while to figure out that `assert()` is calling these implicitly, so if you have an assertion anywhere in your C++ code, that's a problem. I solved it by adding `subdir-ccflags-y += -DNDEBUG` to our app-level `Kbuild` file to force assertions to be removed. I have no idea why this worked in the pre-Kconfig days.

And now I'm left with a single undefined reference:

```
/opt/homebrew/Cellar/gcc-arm-none-eabi/20200630/bin/../lib/gcc/arm-none-eabi/9.3.1/../../../../arm-none-eabi/bin/ld: /opt/homebrew/Cellar/gcc-arm-none-eabi/20200630/bin/../lib/gcc/arm-none-eabi/9.3.1/../../../../arm-none-eabi/lib/thumb/v7e-m+fp/hard/libc.a(lib_a-sbrkr.o): in function `_sbrk_r':
sbrkr.c:(.text._sbrk_r+0xc): undefined reference to `_sbrk'
collect2: error: ld returned 1 exit status
```

It turns out that `sbrk` is a low-level function that is required by `malloc` / `free` / `new` / `delete`, so this is not really related to C++ any more but it's more about dynamic memory allocation by code that is not aware of FreeRTOS and its `pvPortMalloc()` and `vPortFree()` calls. I can work around this by providing a `malloc` / `free` wrapper around the FreeRTOS functions, but I can't figure out why it worked with the pre-Kconfig version.

So, this is the full patch so far (without the `-DNDEBUG` flag because I added that in the app layer):

```patch
diff --git a/src/utils/src/Kbuild b/src/utils/src/Kbuild
index 69a6100b..1b8a5146 100644
--- a/src/utils/src/Kbuild
+++ b/src/utils/src/Kbuild
@@ -5,6 +5,8 @@ obj-y += cpuid.o
 obj-y += crc32.o
 obj-y += debug.o
 obj-y += eprintf.o
+obj-y += abort.o
+obj-y += malloc.o

 ### Add rules for handling generated version.c

diff --git a/src/utils/src/malloc.c b/src/utils/src/malloc.c
new file mode 100644
index 00000000..48cfa7a4
--- /dev/null
+++ b/src/utils/src/malloc.c
@@ -0,0 +1,22 @@
+#include "FreeRTOS.h"
+
+void* malloc(size_t size)
+{
+    void* ptr = NULL;
+
+    if (size > 0)
+    {
+        ptr = pvPortMalloc(size);
+    }
+
+    return ptr;
+}
+
+void free(void* ptr)
+{
+    if (ptr)
+    {
+        vPortFree(ptr);
+    }
+}
+
diff --git a/tools/kbuild/Makefile.kbuild b/tools/kbuild/Makefile.kbuild
index ee2bcc70..92a30fae 100644
--- a/tools/kbuild/Makefile.kbuild
+++ b/tools/kbuild/Makefile.kbuild
@@ -553,7 +553,7 @@ firmware-libs       := $(patsubst %,%/lib.a, $(libs-y))
 firmware-all   := $(firmware-objs) $(firmware-libs)

 quiet_cmd_firmware = LD      $@
-      cmd_firmware = $(LD) $(image_LDFLAGS) $(LDFLAGS) $(firmware-objs) $(firmware-libs) -lm -o $@.elf
+      cmd_firmware = $(LD) $(image_LDFLAGS) $(LDFLAGS) $(firmware-objs) $(firmware-libs) -lstdc++ -lm -o $@.elf

 firmware: $(firmware-all) FORCE
        +$(call if_changed,firmware)
```

I haven't tried uploading the compiled firmware to a drone yet; that's for tomorrow. (Well, today, but after the next sleeping cycle).

Okay, so, first of all, you need `ifneq ($(OOT_USES_CXX),)` in `tools/kbuild/Makefile.kbuild`, not `ifndef`.

Then, the compiler is smart and it knows that the execution can never reach the part after the `while (1)` infinite loop, so anything you put in there is ignored. Moving the construction of `MyClass` above the while loop yields this:

```
/opt/homebrew/Cellar/gcc-arm-none-eabi/20200630/bin/../lib/gcc/arm-none-eabi/9.3.1/../../../../arm-none-eabi/bin/ld: [...]/examples/app_hello_world-cpp/built-in.o: in function `appMain':
[...]/examples/app_hello_world-cpp/src/hello_world.cpp:61: undefined reference to `operator new(unsigned int)'
collect2: error: ld returned 1 exit status
```

Changing to the C++ linker still yields the same error, maybe because of `-nostdlib`? I'm not sure but I then added linking to `stdc++` explicitly (by placing it in front of `-lm`), which gives me this:

```
/opt/homebrew/Cellar/gcc-arm-none-eabi/20200630/bin/../lib/gcc/arm-none-eabi/9.3.1/../../../../arm-none-eabi/bin/ld: /opt/homebrew/Cellar/gcc-arm-none-eabi/20200630/bin/../lib/gcc/arm-none-eabi/9.3.1/../../../../arm-none-eabi/lib/thumb/v7e-m+fp/hard/libc.a(lib_a-abort.o): in function `abort':
abort.c:(.text.abort+0xa): undefined reference to `_exit'
/opt/homebrew/Cellar/gcc-arm-none-eabi/20200630/bin/../lib/gcc/arm-none-eabi/9.3.1/../../../../arm-none-eabi/bin/ld: /opt/homebrew/Cellar/gcc-arm-none-eabi/20200630/bin/../lib/gcc/arm-none-eabi/9.3.1/../../../../arm-none-eabi/lib/thumb/v7e-m+fp/hard/libc.a(lib_a-sbrkr.o): in function `_sbrk_r':
sbrkr.c:(.text._sbrk_r+0xc): undefined reference to `_sbrk'
/opt/homebrew/Cellar/gcc-arm-none-eabi/20200630/bin/../lib/gcc/arm-none-eabi/9.3.1/../../../../arm-none-eabi/bin/ld: /opt/homebrew/Cellar/gcc-arm-none-eabi/20200630/bin/../lib/gcc/arm-none-eabi/9.3.1/../../../../arm-none-eabi/lib/thumb/v7e-m+fp/hard/libc.a(lib_a-signalr.o): in function `_kill_r':
signalr.c:(.text._kill_r+0x12): undefined reference to `_kill'
/opt/homebrew/Cellar/gcc-arm-none-eabi/20200630/bin/../lib/gcc/arm-none-eabi/9.3.1/../../../../arm-none-eabi/bin/ld: /opt/homebrew/Cellar/gcc-arm-none-eabi/20200630/bin/../lib/gcc/arm-none-eabi/9.3.1/../../../../arm-none-eabi/lib/thumb/v7e-m+fp/hard/libc.a(lib_a-signalr.o): in function `_getpid_r':
signalr.c:(.text._getpid_r+0x0): undefined reference to `_getpid'
```

You can see that the error about `operator new` is now gone; the ones above are probably due to the usage of `assert()`, and note that `_sbrk` also appeared, so there you go, I think this pretty much reproduces what I've seen in our more complex setup.

--- a/examples/app_hello_world-cpp/Makefile
+++ b/examples/app_hello_world-cpp/Makefile
@@ -20,4 +20,9 @@ CRAZYFLIE_BASE := ../..
 #
 OOT_CONFIG := $(PWD)/app-config
 
+#
+# Indicate that the out-of-tree build should use a C++ linker
+#
+OOT_USES_CXX := 1
+
 include $(CRAZYFLIE_BASE)/tools/make/oot.mk

--- a/examples/app_hello_world-cpp/src/hello_world.cpp
+++ b/examples/app_hello_world-cpp/src/hello_world.cpp
@@ -7,7 +7,7 @@
  *
  * Crazyflie control firmware
  *
- * Copyright (C) 2019 Bitcraze AB
+ * Copyright (C) 2022 Bitcraze AB
  *
  * This program is free software: you can redistribute it and/or modify
  * it under the terms of the GNU General Public License as published by
@@ -31,6 +31,9 @@
 #include <stdint.h>
 #include <stdbool.h>
 
+#include <cassert>
+#include <string>
+
 extern "C"
 {
   #include "app.h"
@@ -43,9 +46,22 @@ extern "C"
 
 #define DEBUG_MODULE "HELLOWORLD"
 
+class MyClass {
+  public:
+    int myNum;
+    std::string myString;
+};
+
 void appMain()
 {
   DEBUG_PRINT("Waiting for activation ...\n");
+
+  MyClass *cl = new MyClass();
+  DEBUG_PRINT("MyClass has a num: %d\n", cl->myNum);
+
+  /* make sure that the assertion is not simple enough to be optimized away
+   * by the compiler */
+  assert(cl->myNum + cl->myString.size() == 0);
 
   while(1) {
     vTaskDelay(M2T(2000));

--- a/src/utils/src/Kbuild
+++ b/src/utils/src/Kbuild
@@ -6,6 +6,10 @@ obj-y += crc32.o
 obj-y += debug.o
 obj-y += eprintf.o
 
+### Implementations for abort(), malloc() and free(), needed to interact with apps written in C++
+obj-y += abort.o
+obj-y += malloc.o
+
 ### Add rules for handling generated version.c
 
 obj-y += version.o

--- a/src/utils/src/malloc.c
+++ b/src/utils/src/malloc.c
@@ -1,0 +1,22 @@
+#include "FreeRTOS.h"
+
+void* malloc(size_t size)
+{
+    void* ptr = NULL;
+
+    if (size > 0)
+    {
+        ptr = pvPortMalloc(size);
+    }
+
+    return ptr;
+}
+
+void free(void* ptr)
+{
+    if (ptr)
+    {
+        vPortFree(ptr);
+    }
+}
+

--- a/tools/kbuild/Makefile.kbuild
+++ b/tools/kbuild/Makefile.kbuild
@@ -256,6 +256,10 @@ OBJCOPY	= $(CROSS_COMPILE)objcopy
 OBJDUMP	= $(CROSS_COMPILE)objdump
 SIZE    = $(CROSS_COMPILE)size
 
+ifneq ($(OOT_USES_CXX),)
+LD		= $(CROSS_COMPILE)g++
+endif
+
 AWK		= awk
 PERL	= perl
 PYTHON	= python3
@@ -556,6 +560,10 @@ firmware-dirs	:= $(objs-y)
 firmware-objs	:= $(patsubst %,%/built-in.o, $(objs-y))
 firmware-libs	:= $(patsubst %,%/lib.a, $(libs-y))
 firmware-all	:= $(firmware-objs) $(firmware-libs)
+
+ifneq ($(OOT_USES_CXX),)
+	firmware-libs += -lstdc++
+endif
 
 quiet_cmd_firmware = LD      $@
       cmd_firmware = $(LD) $(image_LDFLAGS) $(LDFLAGS) $(firmware-objs) $(firmware-libs) -lm -o $@.elf

--- a/tools/make/oot.mk
+++ b/tools/make/oot.mk
@@ -12,6 +12,10 @@ OOT_CONFIG ?= $(OOT)/oot-config
 
 OOT_ARGS ?= -C $(CRAZYFLIE_BASE) OOT=$(OOT) EXTRA_CFLAGS=$(EXTRA_CFLAGS)
 
+ifneq ($(OOT_USES_CXX),)
+OOT_ARGS += OOT_USES_CXX=1
+endif
+
 .PHONY: all clean
 
 all:

